src/main/java/org/elasticsearch/index/shard/ShardPath.java

```diff
@@ -0,0 +1,173 @@
+        this.path = path;
+        this.indexUUID = indexUUID;
+        this.shardId = shardId;
+        this.shardStatePath = shardStatePath;
+    }
+
+    public Path resolveTranslog() {
+        return path.resolve(TRANSLOG_FOLDER_NAME);
+    }
+
+    public Path resolveIndex() {
+        return path.resolve(INDEX_FOLDER_NAME);
+    }
+
+    public Path getDataPath() {
+        return path;
+    }
+
+    public boolean exists() {
+        return Files.exists(path);
+    }
+
+    public String getIndexUUID() {
+        return indexUUID;
+    }
+
+    public ShardId getShardId() {
+        return shardId;
+    }
+
+    public Path getShardStatePath() {
+        return shardStatePath;
+    }
+
+    /**
+     * This method walks through the nodes shard paths to find the data and state path for the given shard. If multiple
+     * directories with a valid shard state exist the one with the highest version will be used.
+     * <b>Note:</b> this method resolves custom data locations for the shard.
+     */
+    public static ShardPath loadShardPath(ESLogger logger, NodeEnvironment env, ShardId shardId, @IndexSettings Settings indexSettings) throws IOException {
+        final String indexUUID = indexSettings.get(IndexMetaData.SETTING_UUID, IndexMetaData.INDEX_UUID_NA_VALUE);
+        final Path[] paths = env.shardPaths(shardId);
+        Tuple<Path, ShardStateMetaData> metaDataTuple = null;
+        for (Path path : paths) {
+            ShardStateMetaData load = ShardStateMetaData.FORMAT.loadLatestState(logger, path);
+            if (load != null) {
+                if (metaDataTuple == null || metaDataTuple.v2().version < load.version) {
+                    metaDataTuple = new Tuple<>(path, load);
+                    if ((load.indexUUID.equals(indexUUID) || IndexMetaData.INDEX_UUID_NA_VALUE.equals(load.indexUUID)) == false) {
+                        throw new ElasticsearchIllegalStateException("index UUID for shard [" + shardId + "] was: " + load.indexUUID + " excepted: " + indexUUID + " on shard path: " + path);
+                    }
+                }
+            }
+        }
+        if (metaDataTuple == null) {
+            return null;
+        } else {
+            final Path dataPath;
+            final Path statePath = metaDataTuple.v1();
+            if (NodeEnvironment.hasCustomDataPath(indexSettings)) {
+                dataPath = env.resolveCustomLocation(indexSettings, shardId);
+            } else {
+                dataPath = statePath;
+            }
+            logger.debug("{} loaded  data path [{}], state path [{}]", shardId, dataPath, statePath);
+            return new ShardPath(dataPath, statePath, indexUUID, shardId);
+        }
+    }
+
+    // nocommit - we need something more extensible but this does the job for now...
+    public static ShardPath findShardPath(NodeEnvironment env, ShardId shardId, @IndexSettings Settings indexSettings) throws IOException {
```

Comment:
can we name this selectNewPathForShard? , to contrast it from `loadShardPath` (findShardPath sounds to me like go and find an existing shard path).